pkg/ignition/enrich.go

```diff
@@ -0,0 +1,154 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package ignition implements an enricher for Ignition configs.
+package ignition
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/url"
+
+	ignition "github.com/flatcar/ignition/config/v2_3"
+	ignitionTypes "github.com/flatcar/ignition/config/v2_3/types"
+	"github.com/pkg/errors"
+	"k8s.io/utils/ptr"
+
+	"github.com/ionos-cloud/cluster-api-provider-proxmox/pkg/cloudinit"
+)
+
+// Enricher is responsible for enriching the Ignition config with additional data.
+type Enricher struct {
+	BootstrapData []byte
+	Hostname      string
+	InstanceID    string
+	ProviderID    string
+	Network       []cloudinit.NetworkConfigData
+}
+
+// Enrich enriches the Ignition config with additional data.
+func (e *Enricher) Enrich() ([]byte, string, error) {
+	ign, err := e.getEnrichConfig()
+	if err != nil {
+		return nil, "", errors.Wrap(err, "getting enrich config")
+	}
+
+	return buildIgnitionConfig(e.BootstrapData, ign)
+}
+
+func (e *Enricher) getEnrichConfig() (*ignitionTypes.Config, error) {
+	ign := &ignitionTypes.Config{
+		Storage: ignitionTypes.Storage{
+			Files: []ignitionTypes.File{
+				{
+					Node: ignitionTypes.Node{
+						Filesystem: "root",
+						Path:       "/etc/hostname",
+						Overwrite:  ptr.To(true),
+					},
+					FileEmbedded1: ignitionTypes.FileEmbedded1{
+						Mode: ptr.To(0644),
+						Contents: ignitionTypes.FileContents{
+							Source: fmt.Sprintf("data:,%s", e.Hostname),
+						},
+					},
+				},
+				{
+					Node: ignitionTypes.Node{
+						Filesystem: "root",
+						Path:       "/etc/proxmox-env",
+						Overwrite:  ptr.To(true),
+					},
+					FileEmbedded1: ignitionTypes.FileEmbedded1{
+						Mode: ptr.To(420),
+						Contents: ignitionTypes.FileContents{
+							Source: fmt.Sprintf("data:,%s", e.getProxmoxEnvContent()),
+						},
+					},
+				},
+			},
+		},
+		Networkd: ignitionTypes.Networkd{
+			Units: []ignitionTypes.Networkdunit{},
+		},
+		Systemd: ignitionTypes.Systemd{
+			Units: []ignitionTypes.Unit{
+				{
+					Name:   "systemd-resolved.service",
+					Enable: true,
+				},
+			},
+		},
+	}
+
+	// populate networkd units
+	nets, err := RenderNetworkConfigData(cloudinit.BaseCloudInitData{NetworkConfigData: e.Network})
+	if err != nil {
+		return nil, errors.Wrap(err, "rendering networkd units")
+	}
+
+	for name, contents := range nets {
+		ign.Networkd.Units = append(ign.Networkd.Units, ignitionTypes.Networkdunit{
+			Name:     name,
+			Contents: string(contents),
+		})
+	}
+
+	return ign, nil
+}
+
+func (e *Enricher) getProxmoxEnvContent() string {
```

Comment:
Pretty certain we want to double quote escape environment variables we set here.

Reply:
I'm not certain that works 100%. I have to look into that.

Reply:
well, it works ✅ 
you can verify that when we provision a new cluster